src/append.jl

```diff
@@ -0,0 +1,168 @@
+"""
+    Arrow.append(file::String, tbl)
+    tbl |> Arrow.append(io_or_file)
+
+Append any [Tables.jl](https://github.com/JuliaData/Tables.jl)-compatible
+`tbl` to an existing arrow formatted file.
+
+Multiple record batches will be written based on the number of
+`Tables.partitions(tbl)` that are provided; by default, this is just
+one for a given table, but some table sources support automatic
+partitioning. Note you can turn multiple table objects into partitions
+by doing `Tables.partitioner([tbl1, tbl2, ...])`, but note that
+each table must have the exact same `Tables.Schema`.
+
+By default, `Arrow.append` will use multiple threads to write multiple
+record batches simultaneously (e.g. if julia is started with `julia -t 8`
+or the `JULIA_NUM_THREADS` environment variable is set).
+
+Supported keyword arguments to `Arrow.append` include:
+  * `compress`: possible values include `:lz4`, `:zstd`, or your own initialized `LZ4FrameCompressor` or `ZstdCompressor` objects; will cause all buffers in each record batch to use the respective compression encoding
+  * `alignment::Int=8`: specify the number of bytes to align buffers to when written in messages; strongly recommended to only use alignment values of 8 or 64 for modern memory cache line optimization
+  * `dictencode::Bool=false`: whether all columns should use dictionary encoding when being written; to dict encode specific columns, wrap the column/array in `Arrow.DictEncode(col)`
+  * `dictencodenested::Bool=false`: whether nested data type columns should also dict encode nested arrays/buffers; other language implementations [may not support this](https://arrow.apache.org/docs/status.html)
+  * `denseunions::Bool=true`: whether Julia `Vector{<:Union}` arrays should be written using the dense union layout; passing `false` will result in the sparse union layout
+  * `largelists::Bool=false`: causes list column types to be written with Int64 offset arrays; mainly for testing purposes; by default, Int64 offsets will be used only if needed
+  * `maxdepth::Int=$DEFAULT_MAX_DEPTH`: deepest allowed nested serialization level; this is provided by default to prevent accidental infinite recursion with mutually recursive data structures
+  * `ntasks::Int`: number of concurrent threaded tasks to allow while writing input partitions out as arrow record batches; default is no limit; to disable multithreaded writing, pass `ntasks=1`
+"""
+function append end
+
+append(file::String; kw...) = x -> append(file::String, x; kw...)
+
+function append(file::String, tbl;
```

Comment:
we should copy the `Arrow.write` definitions we have that allow `::IO` as well. We should also have a functor form like `append(io_or_file; kw...) = x -> append(io_or_file, x; kw...)

Reply:
The functor form is now present for `append`. It does not yet accept an IO but can be added with a caveat that I wanted to put up.

It seems that we need to overwrite the last empty message in an arrow stream when we append to it, so if we accept an IO, we need to be able to `seek` on it to be able to go back to the point where we got the empty message. And the API will be writing the record batches it appends on to the same IO stream. We will be able to do this on a stream opened on a file, or an IOBuffer, but we will not be able to do it on a network stream for example. Will that be acceptable?

Reply:
Yes, those restrictions seem fine to me and worth mentioning in the docs.

Reply:
done in https://github.com/JuliaData/Arrow.jl/pull/160/commits/00da06d181c9f2d385964cd450315dcbcb5f455b